### PR TITLE
Add codes for generic case of trajectory

### PR DESCRIPTION
**Pull request by chaithyagr**
This PR Handles allowing for undersampling with a given trajectory.
An extension to this, which can be worked on but isnt done here, is to have a large number of such trajectories and later choose randomly for different data

### CODE REVIEW

File: tf_fastmri_data/datasets/non_cartesian.py [hunk 1]
```diff
@@ -24,10 +24,12 @@ def __init__(
             acq_type='radial',
             dcomp=True,
             scale_factor=1e6,
+            trajectory=None,
             **kwargs,
         ):
         self.image_size = image_size
         self.acq_type = acq_type
+        self.traj = trajectory
```
Comment (zaccharieramzi) [inline tf_fastmri_data/datasets/non_cartesian.py:32]: why different names?

**Comment by zaccharieramzi** [comment]
Also (sorry to ask I will verify how to do it automatically), can you increase the version number?

**Comment by chaithyagr** [comment]
Yes

File: tf_fastmri_data/datasets/non_cartesian.py (hunk 1 above)
Comment (chaithyagr) [reply tf_fastmri_data/datasets/non_cartesian.py:32]: Haha, no idea really xD

Comment (zaccharieramzi) [inline tf_fastmri_data/datasets/non_cartesian.py:32]: Then may I suggest

```suggestion
            traj=None,
            **kwargs,
        ):
        self.image_size = image_size
        self.acq_type = acq_type
        self.traj = traj
```

File: tf_fastmri_data/datasets/non_cartesian.py [hunk 2]
```diff
@@ -81,7 +90,12 @@ def preprocessing(self, image, kspace):
                 traj[0],
             )
         traj = tf.repeat(traj, tf.shape(image)[0], axis=0)
-        orig_image_channels = ortho_ifft2d(kspace)
+        orig_image_channels = adjust_image_size(
```
Comment (zaccharieramzi) [inline tf_fastmri_data/datasets/non_cartesian.py:93]: if we do the adjustment here, then let's remove it from the `nufft` function.

Comment (zaccharieramzi) [reply tf_fastmri_data/datasets/non_cartesian.py:93]: But I am not sure exactly where would be the best place for it ...

**Comment by chaithyagr** [comment]
Why is travis still waiting? Did you not disable it @zaccharieramzi ?

File: tf_fastmri_data/datasets/non_cartesian.py (hunk 2 above)
Comment (chaithyagr) [reply tf_fastmri_data/datasets/non_cartesian.py:93]: Hmmm, ill have a look.. I see minute issues in the way gpuNUFFT and kbNUFFT gives results at the moment.

**Comment by zaccharieramzi** [comment]
I did disable it but to no avail on this repo ... don't know why

File: tf_fastmri_data/datasets/non_cartesian.py (hunk 2 above)
Comment (chaithyagr) [reply tf_fastmri_data/datasets/non_cartesian.py:93]: Fixed this now

File: setup.py [hunk 3]
```diff
@@ -9,7 +9,7 @@
 
 setuptools.setup(
     name="tf-fastmri-data",
-    version="0.0.3",
+    version="0.0.5",
```
Comment (zaccharieramzi) [inline setup.py:12]: Why the big jump?

File: tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py [hunk 4]
```diff
@@ -26,13 +26,5 @@ def _crop_for_nufft(image, im_size):
...
-    if image_size is not None:
-        image_adapted = tf.cond(
-            tf.math.greater(shape, image_size[-1]),
-            lambda: _crop_for_nufft(image, image_size),
-            lambda: _pad_for_nufft(image, image_size),
-        )
-    else:
-        image_adapted = image
-    kspace = forward_op(image_adapted, ktraj)
+    kspace = forward_op(image, ktraj)
```
Comment (zaccharieramzi) [inline tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:29]: Can you tell me again why you felt it was better to have the size adjustment inside the dataset rather than the nufft?

**Review by zaccharieramzi — CHANGES_REQUESTED** [review]

File: setup.py (hunk 3 above)
Comment (chaithyagr) [reply setup.py:12]: I think current master is at 0.0.4 right? Did I miss something?

Comment (chaithyagr) [reply setup.py:12]: Oops, I had it at 0.0.4 already, i messed up and did another bump. Will fix that

File: tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py (hunk 4 above)
Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:29]: Actually, I think I am rushing. Let me try to make code cleaner... There seems to be redundacy!

**Comment by chaithyagr** [comment]
I have cleaned up some codes. TO me , it looks like we would need one inside nufft and one in dataset (for reshaping the output image which doesnt pass through NUFFT).
Let me know if I got anything grossly wrong here..

File: tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py (hunk 4 above)
Comment (zaccharieramzi) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: I think in the current state you are just applying `adjust_image_size` twice, also once without the `multicoil` option.

Let's have it only once, and with the `multicoil` option I would say within the nufft function, to keep the status quo

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: We need to apply twice right? Once for the input data and once for the output data?

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: And for output image, its not multicoil really right?

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]:
```python
        orig_image_channels = ortho_ifft2d(kspace)
        image = adjust_image_size(image, self.image_size)
        nc_kspace = nufft(self.nufft_obj, orig_image_channels, traj, self.image_size, multicoil=self.multicoil)
        nc_kspace, image = scale_tensors(nc_kspace, image, scale_factor=self.scale_factor)
```
`orig_image_channels` is per channel image from the k-space. We resize this with `multicoil` inside `nufft`. `image` is reference output image. This isnt `multicoil`. 

>> Let's have it only once, and with the multicoil option I would say within the nufft function, to keep the status quo

I think we do this for `orig_image_channels` . But we need to make sure same happens for `image`

Comment (zaccharieramzi) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: Aaaaah I had indeed read too quickly.
I see now where the confusion is.

Indeed the reference output image is not multicoil. But we don't want to reshape it. The image size thing here is just because of the way tfkbnufft is implemented, which currently requires a fixed image size to be efficient. The `image_size` arguments corresponds to that, which is the in-network image size on which the nufft will operate.

However, in my opinion you are still expected to output an image whose size is the one from the dataset.

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: Well i disagree specifically for Non cartesian case. We dont have any norms yet for that right? 
I dont see a better solution for now than to reshaping all the data. Dont you also do the same in your NCPDNet from what I remember? 
Anyway I think this affects very small percentage of the dataset right?

**Comment by chaithyagr** [comment]
Further, I was thinking of adding a minor check code to make it single coil (by doing a virtual coil combination). Surely not in this PR, but do you think this is the right place for it or that we can just keep that closer to the use case?

File: tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py (hunk 4 above)
Comment (zaccharieramzi) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: We don't have norms, but I figured that it was best to keep the same practice as for the cartesian case, especially since we don't actually need to have an image whose size is adjusted.

In NCPDNet, like I said above, I do the resizing only for the application of the NUFFT, the rest is left to the original image size.

It affects only a very small percentage of the dataset by cropping it, but for most of the dataset it actually pads it, so it does affect a large part of the dataset overall.

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: But wouldnt this cause issue for loss and other metrics? Due to a shape mismatch i.e.

**Comment by zaccharieramzi** [comment]
> Further, I was thinking of adding a minor check code to make it single coil (by doing a virtual coil combination).

To make what single coil ?

File: tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py (hunk 4 above)
Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: Ahhh cool, now I think I get it, `orig_shape` holds the original shape with which the things can still be applied right

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: But as of now `orig_shape` is shape from `kspace`. Are we sure it matches the shape of results?

`orig_shape = tf.ones([tf.shape(kspace)[0]], dtype=tf.int32) * tf.shape(kspace)[-1]`

File: tf_fastmri_data/datasets/non_cartesian.py (hunk 1 above)
Comment (chaithyagr) [reply tf_fastmri_data/datasets/non_cartesian.py]: This is resolved right?

File: tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py (hunk 4 above)
Comment (zaccharieramzi) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: hmm let's see.
Here I used it only for the knee data. So I basically didn't have a problem with the output shape actually because it was always 320x320. So this shape was purely for in-network convolutions (I am basically using the original image shape in order not to perform unnecessary convolutions, but it might be overengineering honestly).

I think indeed for the brain data it's going to be another story because you need to keep the actual output shape.
Also the output shape for the brain data has to be specified by more than just the phase direction size.

This might indeed need some rewriting of the data pipeline and the ncpdnet to handle brain data.
Basically, what we can do is to just use the output shape, and not do resizing in the network (I think the gain is marginal), only crop/pad at the end but not more.

Do you see my points?

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: Wasnt this what we were doing in first place? Basically have codes to adjust all input images to dimensions we need?

Comment (chaithyagr) [reply tf_fastmri_data/preprocessing_utils/fourier/non_cartesian.py:16]: thing with brain data is i have seen 320x320 and 384x384 etc... Making all large just makes things really computationally intensive..

File: tf_fastmri_data/datasets/non_cartesian.py [hunk 5]
```diff
@@ -82,11 +93,16 @@ def preprocessing(self, image, kspace):
...
         orig_image_channels = ortho_ifft2d(kspace)
-        nc_kspace = nufft(self.nufft_obj, orig_image_channels, traj, self.image_size)
+        nc_kspace = nufft(self.nufft_obj, orig_image_channels, traj, self.image_size, multicoil=self.multicoil)
         nc_kspace, image = scale_tensors(nc_kspace, image, scale_factor=self.scale_factor)
         image = image[..., None]
         nc_kspaces_channeled = nc_kspace[..., None]
-        orig_shape = tf.ones([tf.shape(kspace)[0]], dtype=tf.int32) * tf.shape(kspace)[-1]
+        orig_shape = tf.ones([tf.shape(kspace)[0]], dtype=tf.int32) * tf.shape(self.image_size)[-1]
+        if self.crop_image_data:
+            image = adjust_image_size(image, self.image_size)
```
Comment (zaccharieramzi) [inline tf_fastmri_data/datasets/non_cartesian.py:102]: I am wondering if in some cases there are brain images which are less (in some dimension) than `320 x 320`. If so, you will have an output dimension mismatch if you don't specify the output shape, even if you did adjust the image size.

Could you check that?

Comment (chaithyagr) [reply tf_fastmri_data/datasets/non_cartesian.py:102]: Wait no, if the dimension is low, cant we still pad the image? Is this like a PR to `fastmri_recon`? 
My point is that we should use `adjust_image_size` wherever possible. That way, we will crop or pad as needed and not have any issues.. What do you think?

Comment (zaccharieramzi) [reply tf_fastmri_data/datasets/non_cartesian.py:102]: I am slightly less comfortable using padding at the end of the network to fit the expected output image size.
Indeed it brings the question of how exactly to pad because you don't exactly have a 0 background.

It probably isn't a good reason from an MRI point of view but I feel less easy about it.

But even so, you would still need to pass in the output image shape to do padding, so it's not a question for this PR.

Comment (chaithyagr) [reply tf_fastmri_data/datasets/non_cartesian.py:102]: Yes but it does become a question whether i should throw error or warn in case the data shape is less than 320

Comment (zaccharieramzi) [reply tf_fastmri_data/datasets/non_cartesian.py:102]: rather less than the expected output shape in the cross domain net

Comment (chaithyagr) [reply tf_fastmri_data/datasets/non_cartesian.py:102]: Yes... Which is `image_size` . I will add a check and add a warning for now..

**Review by zaccharieramzi — APPROVED** [review]
LGTM
